plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/graph/transform/ElkGraphImporter.java

```diff
@@ -342,73 +341,75 @@ private void importHierarchicalGraph(final ElkNode elkgraph, final LGraph lgraph
+                        || sourceNode.getProperty(LayeredOptions.NO_LAYOUT)
+                        || targetNode.getProperty(LayeredOptions.NO_LAYOUT)) {
+                    continue;
                 }
                 
-                // We add the current node's children if two conditions are met: first, the current node's parent is
-                // null or set to INCLUDE_CHILDREN, and second, the child does not have another layout algorithm
-                // configured
-                HierarchyHandling parentHierarchyHandling = elknode.getParent() == null
-                        ? HierarchyHandling.INCLUDE_CHILDREN
-                        : elknode.getParent().getProperty(LayeredOptions.HIERARCHY_HANDLING);
-                if (parentHierarchyHandling == HierarchyHandling.INCLUDE_CHILDREN) {
-                    for (ElkNode child : elknode.getChildren()) {
-                        boolean usesElkLayered = !child.hasProperty(CoreOptions.ALGORITHM)
-                                || child.getProperty(CoreOptions.ALGORITHM).equals(LayeredOptions.ALGORITHM_ID);
-                        
-                        if (usesElkLayered) {
-                            elknodeQueue.add(child);
-                        }
+                // Check if this edge is an inside self-loop
+                boolean isInsideSelfLoop = elkedge.isSelfloop()
+                        && sourceNode.getProperty(LayeredOptions.INSIDE_SELF_LOOPS_ACTIVATE)
+                        && elkedge.getProperty(LayeredOptions.INSIDE_SELF_LOOPS_YO);
+                
+                // Find the graph the edge will be placed in. Basically, if the edge is an inside
+                // self loop or connects one of its end points to a descendant, the edge will be
+                // placed in the graph that represents that end point's insides. Otherwise, it will
+                // be placed in the current graph.
+                ElkNode parentElkGraph = elkGraphNode;
+                
+                if (isInsideSelfLoop || ElkGraphUtil.isDescendant(targetNode, sourceNode)) {
+                    parentElkGraph = sourceNode;
+                } else if (ElkGraphUtil.isDescendant(sourceNode, targetNode)) {
+                    parentElkGraph = targetNode;
+                }
+                
+                LGraph parentLGraph = lgraph;
+                LNode parentLNode = (LNode) nodeAndPortMap.get(parentElkGraph);
+                if (parentLNode != null) {
+                    parentLGraph = parentLNode.getNestedGraph();
+                }
+                
+                // Transform the edge, finally...
+                LEdge ledge = transformEdge(elkedge, parentElkGraph, parentLGraph);
+                
+                // Find the graph the edge's coordinates will have to be made relative to during export
+                LGraph coordinateSystemOrigin = findCoordinateSystemOrigin(elkedge, elkgraph, lgraph);
+                if (coordinateSystemOrigin != null) {
```

Comment:
If it's `null`, it's assumed to be already in the correct coordinate system?

Reply:
Yep, but that's an unnecessary distinction, I guess. If it's `null`, the `setProperty(...)` call wouldn't change anything anyway, so I'll just remove the conditional statement there.